### PR TITLE
[TIC 110] Strictly require the Complete Order contact form

### PR DESCRIPTION
**Pull request by benfaustin95**
### Description
Added form input validation for checkout form. 
Added pop up on checkout failure. 

### Risks
Incorrect logic could have been used for input validation. 

### Validation
Manual testing 

### Issue
NA

### Operating System
M1 Mac

### CODE REVIEW

File: client/src/components/Ticketing/event/TicketPicker.tsx [hunk 1]
```diff
@@ -392,12 +392,11 @@ const TicketPicker = (props: TicketPickerProps): ReactElement => {
               className='text-zinc-300'
               value=''
               disabled
```
Comment (henrykaus) [inline client/src/components/Ticketing/event/TicketPicker.tsx:394]: Thanks for fixing all of the console errors in this file!! They were bothering me but I didn't exactly know which things to remove

**Review by henrykaus — COMMENTED** [review]
I noticed a problem when testing out this branch. I found that if you were to fill out all of the fields (removing all of the red required messages), then empty a field or 2, then the validation wouldn't occur and you'd still make it to the checkout page.

File: client/src/components/Ticketing/checkout/CompleteOrderForm.tsx [hunk 2]
```diff
@@ -80,293 +61,220 @@ export default function CompleteOrderForm({
...
+    const errors = {};
+    Object.keys(values).forEach((key) => {
+      if (['visitSource', 'comments', 'optIn'].includes(key)) return;
+      if (!values[key] || values[key] === '') {
+        errors[key] = 'Required';
+      }
+    });
+    if (values.seatingAcc === 'Other' && (!values.comments || values.comments === '')) {
+      errors['comments'] = 'Please Input Accommodation';
+    }
```
Comment (henrykaus) [inline client/src/components/Ticketing/checkout/CompleteOrderForm.tsx:89]: This is a nice feature!

File: client/src/components/Ticketing/checkout/CheckoutFormInput.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+export const FormInput = (props: FormInputProps) => {
+  const {
+    id,
+    name,
+    type,
+    input,
+    inputClass,
+    labelClass,
+    divClass,
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CheckoutFormInput.tsx:24]: Should I do an object prop called className with input, label, and div values?

[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CheckoutFormInput.tsx]: Fixed

File: client/src/components/Ticketing/checkout/CheckoutPage.tsx [hunk 4]
```diff
@@ -68,64 +77,87 @@ export default function CheckoutPage(): ReactElement {
...
+              viewBox='0 0 20 20'
+              fill='currentColor'
+            >
+              <path
+                fillRule='evenodd'
+                d='M9.707 14.707a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 1.414L7.414 9H15a1 1 0 110 2H7.414l2.293 2.293a1 1 0 010 1.414z'
+                clipRule='evenodd'
+              />
+            </svg>
           back to Events
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CheckoutPage.tsx]: fixed

File: client/src/components/Ticketing/checkout/CheckoutPage.tsx [hunk 5]
```diff
@@ -68,64 +77,87 @@ export default function CheckoutPage(): ReactElement {
...
+          </button>
+        </div>
+        <div className='flex flex-row items-center mt-2 text-zinc-800'>
+          <div className='text-4xl font-bold'>Checkout</div>
+        </div>
+        <div className='flex flex-col items-center md:flex-row md:items-stretch sm:flex-col w-full h-full'>
+          <div className='min-w-414 sm:w-full h-full md:mt-10 sm:mt-10 bg-zinc-100 p-2 pt-4 md:p-9 flex flex-col gap-5 items-start rounded-xl overflow-auto'>
+            <div className='flex flex-col items-center h-auto w-full'>
+              <div className='text-2xl lg:text-5xl font-bold mb-5'>
               Complete Order
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CheckoutPage.tsx]: fixed

File: client/src/components/Ticketing/checkout/CheckoutFormInput.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+
+interface FormInputProps {
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CheckoutFormInput.tsx]: fixed

File: client/src/components/Ticketing/checkout/CompleteOrderForm.tsx [hunk 7]
```diff
@@ -80,293 +61,220 @@ export default function CompleteOrderForm({
...
+                    <Field
+                      component='input'
+                      type='checkbox'
+                      name='optIn'
+                    />
+                    <div>
                         I would like to receive email info from Portland
                         Playhouse
-                      </div>
                     </div>
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CompleteOrderForm.tsx]: fixed

File: client/src/components/Ticketing/checkout/CompleteOrderForm.tsx [hunk 8]
```diff
@@ -80,293 +61,220 @@ export default function CompleteOrderForm({
...
+                    type='text'
+                    placeholder='What do you want us to know?'
+                    id='comments'
+                    label='Comments'
+                    labelClass='block text-sm font-medium text-slate-700 ml-1'
+                    inputClass='input w-full border border-zinc-300 p-4 rounded-lg'
+                  />
+                </div>
+                <div className='flex flex-col items-start gap-3 mt-5 mb-5'>
+                  <div className='flex flex-row items-center gap-4 text-sm text-zinc-700'>
```
[the remark this answers is not in the document]
Comment (benfaustin95) [reply client/src/components/Ticketing/checkout/CompleteOrderForm.tsx]: fixed

File: client/src/components/Ticketing/checkout/CheckoutPage.tsx [hunk 9]
```diff
@@ -38,6 +39,14 @@ export default function CheckoutPage(): ReactElement {
...
     'donation',
   );
+  const [popUp, setPopUp] = useState({
+    show: false,
+    title: '',
+    message: '',
+    success: true,
+    handleProceed: () => setPopUp((popUp) => ({...popUp, show: false})),
+    showSecondary: false,
+  });
```
Comment (henrykaus) [inline client/src/components/Ticketing/checkout/CheckoutPage.tsx:49]: I noticed that the `x` button of the PopUp doesn't work, its likely that you'll need to add a `handleClose: () => setPopUp((popUp) => ({...popUp, show: false})),`, and with this observation, we probably don't want the `handleClose` function to be optional in the `PopUp` file. That ones on me

**Review by henrykaus — APPROVED** [review]

**Pull request merged**